vernac/comAssumption.ml

```diff
@@ -72,211 +103,139 @@ let interp_assumption ~program_mode env sigma impl_env bl c =
-     restrict_universe_context should be called in a centralized place
-     IMO, thus I think we should adapt `prepare_parameter` to handle
-     this case too. *)
-  let sigma = Evd.restrict_universe_context sigma uvars in
-  let univs = Evd.check_univ_decl ~poly sigma udecl in
-  declare_assumptions ~scope ~kind ?user_warns univs nl l
-
-let context_subst subst (name,b,t,impl) =
-  name, Option.map (Vars.substl subst) b, Vars.substl subst t, impl
+  | _, (UState.Monomorphic_entry _, _) -> empty_univ_entry false
+  | _, (UState.Polymorphic_entry _, _) -> empty_univ_entry true
 
-let context_insection sigma ~poly ctx =
-  let uctx = Evd.evar_universe_context sigma in
-  let univs = UState.univ_entry ~poly uctx in
-  let fn i subst (name,_,_,_ as d) =
-    let d = context_subst subst d in
-    let univs = if i = 0 then univs else empty_univ_entry ~poly in
-    let refu = match d with
-      | name, None, t, impl ->
-        let kind = Decls.Context in
-        declare_variable NoCoercion ~kind t univs [] impl name
-      | name, Some b, t, impl ->
-        let entry = Declare.definition_entry ~univs ~types:t b in
-        (* XXX Fixme: Use Declare.prepare_definition *)
-        let kind = Decls.(IsDefinition Definition) in
-        let gr =
-          Declare.declare_entry ~name ~scope:Locality.Discharge
-            ~kind ~impargs:[] ~uctx entry
-        in
-        (gr,UVars.Instance.empty)
-    in
-    Constr.mkRef refu :: subst
-  in
-  let _ : Vars.substl = List.fold_left_i fn 0 [] ctx in
-  ()
+let context_subst subst (id,b,t,infos) =
+  id, Option.map (Vars.replace_vars subst) b, Vars.replace_vars subst t, infos
 
-let context_nosection sigma ~poly ctx =
-  let (univ_entry,ubinders as univs) = Evd.univ_entry ~poly sigma in
+let declare_context ~try_global_assum_as_instance ~scope ~univs ?user_warns ~inline ctx =
   let fn i subst d =
-    let (name,b,t,_impl) = context_subst subst d in
-    let kind = Decls.(IsAssumption Logical) in
-    let local = if Lib.is_modtype () then Locality.ImportDefaultBehavior
-      else Locality.ImportNeedQualified
-    in
-    (* Multiple monomorphic axioms: declare universes only on the first declaration *)
-    let univs = if i = 0 then univs else clear_univs (Locality.Global local) univs in
-    let decl = match b with
-      | None ->
-        let entry = Declare.parameter_entry ~univs:(univ_entry, ubinders) t in
-        Declare.ParameterEntry entry
-      | Some b ->
-        let entry = Declare.definition_entry ~univs ~types:t b in
-        Declare.DefinitionEntry entry
-    in
-    let cst = Declare.declare_constant ~name ~kind ~local decl in
-    let () = Declare.assumption_message name in
-    let env = Global.env () in
-    let () = if Option.is_empty b then
-        Classes.declare_instance env sigma None Hints.SuperGlobal (GlobRef.ConstRef cst)
-    in
-    Constr.mkConstU (cst,instance_of_univ_entry univ_entry) :: subst
+    let (name,b,t,(impl,kind,coe,impargs)) = context_subst subst d in
```

Comment:
Note that (not in this PR), those would probably benefit from using records rather than tuples of tuples.

Reply:
OK